src/provider/index.ts

```diff
@@ -163,11 +164,45 @@ export class Main {
      * @param sender The source of the notification.
      */
     private hydrateNotification(payload: NotificationOptions, sender: Identity): StoredNotification {
-        if (!payload.body) {
-            throw new Error('Invalid arguments passed to createNotification. "body" must have a value');
+        const problems: string[] = [];
+
+        if (payload.id !== undefined && typeof payload.id !== 'string') {
+            problems.push('"id" must be a string or undefined');
+        }
+
+        if (typeof payload.body === 'undefined') {
+            problems.push('"body" must have a value');
+        } else if (typeof payload.body !== 'string') {
+            problems.push('"body" must be a string');
+        }
+
+        if (typeof payload.title === 'undefined') {
+            problems.push('"title" must have a value');
+        } else if (typeof payload.title !== 'string') {
+            problems.push('"title" must be a string');
+        }
+
+        if (payload.subtitle !== undefined && typeof payload.subtitle !== 'string') {
+            problems.push('"subtitle" must be a string or undefined');
         }
-        if (!payload.title) {
-            throw new Error('Invalid arguments passed to createNotification. "title" must have a value');
+
+        if (payload.icon !== undefined && typeof payload.icon !== 'string') {
+            problems.push('"icon" must be a string or undefined');
+        }
+
+        const parsedDate = moment(payload.date);
+        if (payload.date !== undefined && !parsedDate.isValid()) {
+            problems.push('"date" must be a date string, unix timestamp, or undefined');
```

Comment:
A `Date` object also ought to be acceptable - or are those `toString`-d when they come over the channel? The error message should probably reflect what we expect applications to pass to the client though, rather than what we expect the provider to receive from the client.

Also, this range of inputs sounds like ideal behaviour, but isn't how we type it on the client-side. Do we need to make `date` into a `Date|string|number`?.. Does make the definition of `Notification` slightly more fiddly, though.

Reply:
Yea, I struggled a bit with this one. It's a bit awkward, because as you guessed, the dates are stringified over the bus, so while we get a "date-like" object it is not actually an instance of the date class. I think we'd be best keeping the client-side as `Date`, and update the error message on this side to reflect that

Reply:
As for the second point there is no technical reason why we couldn't accept date strings in the client, but it wouldn't be possible to fully typeguard them. At least by leaving it as only accepting `Date` the users can be confident they're passing the correct data at compile time.